### PR TITLE
Enumerable methods

### PR DESCRIPTION
**Pull request by carloso0114**
This project consisted of re-building some of the most used enumerable methods in Ruby: each, my_each_with_inde, my_select, my_all? my_any? my_none? my_count, my_map, my_inject, and an extra method called 'multiply_els'.
For this project, we used Ruby, Git, and Rubocop.
For documentation and learning, we used the ruby-docs site, StackOverflow, and Codecademy.

****Note:** This will be our second code review, this repository was created because we had branch conflicts with the first one, so we decided to create a new one fixing the details and errors that the first code reviewer noticed.
Previous repository: https://github.com/carloso0114/enumerable_methods/pull/2

### CODE REVIEW

File: e_methods.rb [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+      my_each { |value| result.push(value) if yield(value) }
+    else
+      result = {}
+      my_each { |key, value| result[key] = value if yield(key, value) }
+    end
+    result
+  end
+
+  def my_all?(argum = nil)
+    if block_given? && !argum.nil?
```
Comment (kevcap) [inline e_methods.rb:46]: #my_all? when a pattern other than Regex or a Class is given returns false if any of the collection does not match the pattern
     Failure/Error: expect(array.my_all?(3)).to be array.all?(3)

       expected false
            got true

File: e_methods.rb [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+    elsif !argum.nil? && argum.instance_of?(Regexp)
+      my_each { |element| return false unless argum.match(element) }
+    else
+      my_each { |element| return false unless element }
+    end
+    true
+  end
+
+  def my_any?(argum = nil)
+    if block_given? && !argum.nil?
```
Comment (kevcap) [inline e_methods.rb]: when no block or argument is given returns false if at least one of
the collection is not true
     Failure/Error: expect(false_array.my_any?).to be false_array.any?

       expected false
            got true

File: e_methods.rb [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+    elsif argum
+      my_each { |element| return true if element == argum }
+    else
+      my_each { |element| return true if element }
+    end
+    false
+  end
+
+  def my_none?(argum = nil)
+    if block_given? && !argum.nil?
```
Comment (kevcap) [inline e_methods.rb:90]: 1- returns false if the block ever returns true for all elements::range
     Failure/Error: expect(range.my_none?(&false_block)).to eq(range.none?(&false_block))
 ```
      expected: false
            got: true
```
      note: It was compared using ==;

2- when no block or argument is given returns false only if one of the collection members is true
     Failure/Error: expect(true_array.my_none?).to be false

       expected false
            got true

3- when a pattern other than Regex or a Class is given returns false only if one of the collection matches the pattern
     Failure/Error: expect(words.my_none?(5)).to be words.none?(5)

       expected false
            got true

File: e_methods.rb [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+    if !prc.nil?
+      to_a.my_each { |i| new_arr.push(prc.call(i)) }
+    else
+      to_a.my_each { |i| new_arr.push(yield(i)) }
+    end
+    new_arr
+  end
+
+  def my_inject(argum = nil, sym = nil)
+    if !block_given?
```
Comment (kevcap) [inline e_methods.rb:140]: Should raise a "LocalJumpError" when no block or argument is given
     Failure/Error: expect { array.my_inject }.to raise_error(LocalJumpError)
       expected LocalJumpError but nothing was raised

**Review by kevcap — CHANGES_REQUESTED** [review]
# [STATUS] CHANGES REQUIRED♻️

#### Hey team,  your project is looking really good, however, to accomplish as the actual project, you required to do these changes, listed below:
 
 
## [REQUIRED CHANGES]
- [x] Note that your methods should work exactly as the built-in ruby methods, you can find all of them here: https://ruby-doc.org/core-2.7.1/
- [x] I've made some comments down below, please check what was expected and what was actually given, with that info you guys can search what was the issue and how to solve it;

## Please make the required changes and resubmit for a new code review once you are ready💪

## As always, I wish you all the very best and see you both at the top🚀


👨🏻‍💻 **TSE: Kevin Alves**

  

📫 Have any question? Better call Kevin:

<a  href="mailto:kevinalves.js@gmail.com"> gmail</a> [Twitter Follow](https://twitter.com/intent/follow?screen_name=kevinnnmn) [](https://microverse-students.slack.com/team/UT86D3AP4)

<sub>*We're here to help you, to ensure that you're using the best practices and getting you ready for the job market. In real-world, you'll have a lot of changes requests for your real clients, it's good to get used to it.</sub>

File: e_methods.rb [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+      my_each { |element| return false unless element.is_a? argum }
+    elsif !argum.nil? && argum.instance_of?(Regexp)
+      my_each { |element| return false unless argum.match(element) }
+    elsif !argum.nil?
+      my_each { |element| return false unless element == argum }
+    else
+      my_each { |element| return false unless element }
+    end
+    true
+  end
```
Comment (elmejdki) [inline e_methods.rb:63]: ## my_all?:

- [x] when a pattern other than Regex or a Class is given returns false if any of the collection does not match the pattern:

```ruby

    array = [23, 43, 56, 2, 64, 76, 30, 16, 43, 26]
    array.my_all?(3) === array.all?(3) => false

```

File: e_methods.rb [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+    elsif !argum.nil? && argum.instance_of?(Regexp)
+      my_each { |element| return true unless argum.match(element) }
+    elsif argum
+      my_each { |element| return false if element == argum }
+      my_each { |element| return true if element != argum }
+    else
+      my_each { |element| return true unless element }
+    end
+    false
+  end
```
Comment (elmejdki) [inline e_methods.rb:111]: ## my_any?:

- [x]  returns false if the block never returns true for all elements::range:

```ruby

      range = Range.new(5, 50)
      false_block = proc { |num| num > 51 }
      range.my_none?(&false_block) === range.none?(&false_block) => false

```

File: e_methods.rb [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+    elsif block_given? && argum
+      memo = argum
+      my_each { |element| memo = yield(memo, element) }
+    else
+      memo = first
+      to_a[1..to_a.length].my_each { |element| memo = yield(memo, element) }
+      memo
+    end
+    memo
+  end
```
Comment (elmejdki) [inline e_methods.rb]: ## my_inject:

and this [review](https://github.com/carloso0114/ruby_enumerators/pull/1/files#r536390674) you have forgotten to work on, it's super clear I think.

`
Should raise a "LocalJumpError" when no block or argument is given
Failure/Error: expect { array.my_inject }.to raise_error(LocalJumpError)
expected LocalJumpError but nothing was raised
`

**Review by elmejdki — CHANGES_REQUESTED** [review]
# status: require changes :recycle: 

Impressive work @team, you are almost there, you have actually left some of the required changes by the previous TSE don't worry I will give you more details so you know how to solve them.

## Required changes:

Please check the comments below :point_down: :point_down: :point_down: :point_down: :point_down: 

### Please make all the required changes before you submit another code review.
## Happy Coding :tophat: 
_Feel free to contact me on slack if you have any questions._

**Review by elmejdki — APPROVED** [review]
# status: approved :heavy_check_mark: 

Great job guys :100: :1st_place_medal: , thanks for making the required changes:+1:.

now you can merge to development you project is approved.

## Happy Coding :tophat:

**Pull request merged**